### PR TITLE
Better handling of let's

### PR DESCRIPTION
**Pull request by fblanqui**
partialy solve #648

### CODE REVIEW

**Comment by fblanqui** [comment]
@gabrielhdt Let me know if you want to have a look at this PR.

File: src/core/eval.ml [hunk 1]
```diff
@@ -474,19 +447,44 @@ let unfold_sym : sym -> term -> term =
   match !(s.sym_rules) with
   | [] -> fun t -> t
   | _ ->
-      let dt = !(s.sym_dtree) in
+      let c = cfg_of_ctx [] true and dt = !(s.sym_dtree) in
```
Comment (gabrielhdt) [inline src/core/eval.ml:450]: The `and` is not needed here, if I'm not mistaken.

Comment (fblanqui) [reply src/core/eval.ml:450]: What do you mean? This just means that the two can expressions be computed in parallel.

Comment (gabrielhdt) [reply src/core/eval.ml:450]: Yes, but it introduces unneeded mutual dependencies between definitions, and doesn't speed up anything.

Comment (fblanqui) [reply src/core/eval.ml:450]: No, its a let not a let rec.

Comment (gabrielhdt) [reply src/core/eval.ml:450]: Ah right!

File: src/core/infer.ml [hunk 2]
```diff
@@ -136,74 +126,88 @@ let rec infer : problem -> ctxt -> term -> term = fun p ctx t ->
...
+      | Kind ->
+          wrn None "Abstraction on [%a] is not allowed." Print.pp_term u;
+          raise NotTypable
+      | _ ->
+          let s = infer p c' b in
+          begin match unfold s with
+          | Type | Kind -> ()
+          | _ -> conv p c' s mk_Type
+          (* Here, we force [s] to be equivalent to [Type] as there is little
+             (no?) chance that it can be a kind. *)
```
[the remark this answers is not in the document]
Comment (gabrielhdt) [reply src/core/infer.ml:162]: But this involves a lot of modifications, it's just a sidenote

File: src/core/eval.ml [hunk 3]
```diff
@@ -40,122 +40,130 @@ let eta_equality : bool ref = Console.register_flag "eta_equality" false
...
-  (* In head beta normal form. *)
-  | (_        , _      ) -> st
-
-(** [whnf_beta t] computes a weak head beta normal form of [t]. *)
-let whnf_beta : term -> term = fun t ->
-  Stdlib.(steps := 0);
-  let u = whnf_beta t in
-  if Stdlib.(!steps = 0) then unfold t else u
+(** [snf whnf t] computes a snf of [t] using [whnf]. *)
+let snf : (term -> term) -> (term -> term) = fun whnf ->
```
[the remark this answers is not in the document]
Comment (fblanqui) [reply src/core/eval.ml:54]: Right but it helps figure out more quickly that it is a functional on `term -> term`.

File: src/core/infer.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (fblanqui) [reply src/core/infer.ml:162]: Note that this is a copy and paste of the Prod case. It doesn't seem necessary at the moment. That's better if we can avoid backtracking.

[the remark this answers is not in the document]
Comment (gabrielhdt) [reply src/core/infer.ml:162]: I guess we can't avoid it, in some cases (it isn't needed for the moment indeed).

File: src/core/eval.ml [hunk 4]
```diff
@@ -40,122 +40,130 @@ let eta_equality : bool ref = Console.register_flag "eta_equality" false
...
-  if !log_enabled && not res then log_conv "failed"; res
+  fun a b ->
+  if !log_enabled then log_conv "%a ≟ %a" pp_term a pp_term b;
+  try eq_modulo [(a,b)]; true
+  with Exit -> if !log_enabled then log_conv "failed"; false
+
+(** [Configuration of the reduction engine. *)
+type config =
+  {context : ctxt; (** Context of the reduction used for generating metas. *)
+   defmap : term VarMap.t; (** Variable definitions. *)
```
[the remark this answers is not in the document]
Comment (fblanqui) [reply src/core/eval.ml]: I don't have serious tests to measure performance. On `make tests`, I get a gain of .5s (this is what we lost with #687).
Why do you think that it is not worth? The point is that, in whnf_stk line 164, we are checking variable definitions.

File: src/core/infer.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (fblanqui) [reply src/core/infer.ml:162]: It would be nice to have some example.

**Comment by fblanqui** [comment]
@gabrielhdt I fixed most of your remarks. Let me know if you are happy.

File: src/core/eval.ml (hunk 4 above)
[the remark this answers is not in the document]
Comment (gabrielhdt) [reply src/core/eval.ml]: Just to be sure that such modification is worth something, I mean that we don't want to alter the code (and complexify datatypes) for nothing.

[the remark this answers is not in the document]
Comment (fblanqui) [reply src/core/eval.ml]: Sure. Not that this config data type has other advantages. In fact, we do not need defmap to correspond to context. An empty map would disable delta-reduction. Later, we could also restrict rewriting to a given set of symbols. Etc.

File: src/core/eval.ml [hunk 5]
```diff
@@ -40,122 +40,133 @@ let eta_equality : bool ref = Console.register_flag "eta_equality" false
...
+      begin match !(s.sym_def) with
+      | Some t when not s.sym_opaq -> Stdlib.incr steps; whnf_stk c t stk
+      | _ ->
+      match tree_walk c !(s.sym_dtree) stk with
+      | None -> r
+      | Some(t,stk) -> Stdlib.incr steps; whnf_stk c t stk
+      end
+  | (Vari x, stk) as r ->
+      begin match VarMap.find_opt x c.defmap with
+      | Some v -> Stdlib.incr steps; whnf_stk c v stk
```
Comment (fblanqui) [inline src/core/eval.ml:166]: @gabrielhdt Note that the normalization of definitions is not shared. We may want to improve that too.

File: src/core/eval.ml [hunk 6]
```diff
@@ -40,122 +40,133 @@ let eta_equality : bool ref = Console.register_flag "eta_equality" false
...
+and whnf_stk : config -> term -> stack -> term * stack = fun c t stk ->
+  (*if !log_enabled then
+    log_eval "whnf_stk %a %a"
+      pp_ctxt c.context pp_term t (D.list pp_term) stk;*)
+  match unfold t, stk with
+  | Appl(f,u), stk -> whnf_stk c f (appl_to_tref u::stk)
+  | Abst(_,f), u::stk ->
+      Stdlib.incr steps; whnf_stk c (Bindlib.subst f u) stk
+  | LLet(_,t,u), stk ->
+      Stdlib.incr steps; whnf_stk c (Bindlib.subst u (appl_to_tref t)) stk
```
Comment (fblanqui) [inline src/core/eval.ml:155]: @gabrielhdt Here, we could extend c.defmap instead. I don't know what would be the best.

**Pull request merged**